src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java

```diff
@@ -438,22 +445,24 @@ public String getTopicDirFromTopic(String topic) {
       }
 
       // make sure that all references to captured groups actually exist
-      int maxVar = extractMaximumInt(topicsDir);
-      if (maxVar > topicGroups.size()) {
-        throw new ConfigException(
-            String.format(
-                "Topic %s must be split into at least %d parts using regex pattern %s, "
-                    + "but was actually split into %d parts",
-                topic,
-                maxVar,
-                topixRegexCaptureGroup.pattern(),
-                topicGroups.size()
-            )
-        );
+      if (!partsToReplace.isEmpty()) {
+        int largestVar = partsToReplace.get(partsToReplace.size() - 1);
+        if (largestVar > topicGroups.size()) {
+          throw new ConfigException(
+              String.format(
+                  "Topic %s must be split into at least %d parts using regex pattern %s, "
+                      + "but was actually split into %d parts",
+                  topic,
+                  largestVar,
+                  topixRegexCaptureGroup.pattern(),
+                  topicGroups.size()
+              )
+          );
+        }
       }
 
-      for (int i = 0; i < maxVar; i++) {
-        topicsDir = topicsDir.replace("${" + (i + 1) + "}", topicGroups.get(i));
+      for (int index : partsToReplace) {
+        topicsDir = topicsDir.replace("${" + index + "}", topicGroups.get(index - 1));
```

Comment:
Another way to do this that might be even nicer is to make `topicGroups` a Map<String, String> mapping "${1}" -> "captured_value_1", "${2}" -> "captured_value_2", etc. Then you can do something like 
```
topicGroups.forEach((k, v) -> topicsDir.replaceAll(k, v));
```
If you still see something that looks like "${\d+}" after that, user tried to reference invalid capture group, throw.

Reply:
I changed the implementation + behavior slightly (see one pager for full explanation) so the code is slightly different not, but let me know what you think